File: haskell/src/Vregex.hs
-- Minimal backtracking regex engine for the Haskell port of voxgig/struct.
-- Supports the RE2 subset the corpus exercises: literals, '.', anchors ^ $,
-- \b and \B, character classes [..] / [^..] with ranges and \d \w \s \D \W \S,
-- groups (..), (?:..) and (?P<name>..), alternation |, quantifiers * + ? and
-- {n}/{n,}/{n,m} with optional lazy '?'. Capturing groups are tracked, so the
-- engine backs the full re_* API (find with captures, find_all, replace) at
-- the Go-stdlib minimum bar; `test` also serves $LIKE. No third-party
-- dependency.

module Vregex
  ( Re
  , compile
  , test
  , testStr
  , findBounds
  , find
  , findAll
  , replaceAll
  ) where

import Control.Applicative ((<|>))
import Data.Array (Array, listArray, (!))
import Data.Foldable (asum)
import Data.Maybe (isJust)

data Node
  = Char Char
  | Any
  | Start
  | End
  | WordB
  | NWordB
  | Cls Bool [Citem]              -- negated?, items
  | Grp Int [[Node]]              -- capture index (0 = non-capturing), alternation
  | Star Bool Node               -- greedy?, atom
  | Plus Bool Node
  | Opt Bool Node
  | Rep Bool Int (Maybe Int) Node

data Citem
  = CChar Char
  | CRange Char Char
  | CD | CW | CS | CND | CNW | CNS  -- \d \w \s \D \W \S

-- Compiled = the alternation AST plus the number of capturing groups.
data Re = Re { reAlts :: [[Node]], reNgroups :: Int }

-- ----- parser (remaining-string style, threading the group counter) -----

parse :: String -> Re
parse pat = let (alts, gc, _) = parseAlt 0 pat in Re alts gc

parseAlt :: Int -> String -> ([[Node]], Int, String)
parseAlt gc0 s0 =
  let (first, gc1, s1) = parseSeq gc0 s0
  in go [first] gc1 s1
  where
    go acc gc ('|':rest) = let (sq, gc', r) = parseSeq gc rest in go (sq : acc) gc' r
    go acc gc r = (reverse acc, gc, r)

parseSeq :: Int -> String -> ([Node], Int, String)
parseSeq = goSeq []
  where
    goSeq acc gc s = case s of
      [] -> (reverse acc, gc, s)
      ('|':_) -> (reverse acc, gc, s)
      (')':_) -> (reverse acc, gc, s)
      _ -> case parseAtom gc s of
             (Nothing, gc', s') -> (reverse acc, gc', s')
             (Just a, gc', s') ->
               let (a', s'') = parseQuantSuffix a s'
               in goSeq (a' : acc) gc' s''

parseAtom :: Int -> String -> (Maybe Node, Int, String)
parseAtom gc s = case s of
  [] -> (Nothing, gc, s)
  ('(':rest) ->
    -- (?: is non-capturing; a plain ( and an RE2 named group (?P<name> are
    -- both capturing (names are not tracked). Indices follow open parens.
    let (gi, gc1, rest1) = case rest of
          ('?':':':r) -> (0, gc, r)
          ('?':'P':'<':r) ->
            let r' = drop 1 (dropWhile (/= '>') r)
            in (gc + 1, gc + 1, r')
          _ -> (gc + 1, gc + 1, rest)
        (alts, gc2, r2) = parseAlt gc1 rest1
        r3 = case r2 of (')':r) -> r; _ -> r2
    in (Just (Grp gi alts), gc2, r3)
  ('[':_) -> let (n, r) = parseClass s in (Just n, gc, r)
  ('.':rest) -> (Just Any, gc, rest)
  ('^':rest) -> (Just Start, gc, rest)
  ('$':rest) -> (Just End, gc, rest)
  ('\\':rest) -> case rest of
    ('d':r) -> (Just (Cls False [CD]), gc, r)
    ('w':r) -> (Just (Cls False [CW]), gc, r)
    ('s':r) -> (Just (Cls False [CS]), gc, r)
    ('D':r) -> (Just (Cls False [CND]), gc, r)
    ('W':r) -> (Just (Cls False [CNW]), gc, r)
    ('S':r) -> (Just (Cls False [CNS]), gc, r)
    ('b':r) -> (Just WordB, gc, r)
    ('B':r) -> (Just NWordB, gc, r)
    ('n':r) -> (Just (Char '\n'), gc, r)
    ('t':r) -> (Just (Char '\t'), gc, r)
    ('r':r) -> (Just (Char '\r'), gc, r)
    (c:r) -> (Just (Char c), gc, r)
    [] -> (Just (Char '\\'), gc, [])
  (c:rest) -> (Just (Char c), gc, rest)

parseClass :: String -> (Node, String)
parseClass ('[':s0) =
  let (neg, s1) = case s0 of ('^':r) -> (True, r); _ -> (False, s0)
      (items, rest) = goCls [] s1
  in (Cls neg (reverse items), rest)
  where
    goCls acc s = case s of
      [] -> (acc, s)
      (']':r) -> (acc, r)
      ('\\':r) -> case r of
        ('d':r') -> goCls (CD : acc) r'
        ('w':r') -> goCls (CW : acc) r'
        ('s':r') -> goCls (CS : acc) r'
        ('D':r') -> goCls (CND : acc) r'
        ('W':r') -> goCls (CNW : acc) r'
        ('S':r') -> goCls (CNS : acc) r'
        ('n':r') -> goCls (CChar '\n' : acc) r'
        ('t':r') -> goCls (CChar '\t' : acc) r'
        ('r':r') -> goCls (CChar '\r' : acc) r'
        (c:r') -> goCls (CChar c : acc) r'
        [] -> (acc, [])
      (c : '-' : c2 : r) | c2 /= ']' -> goCls (CRange c c2 : acc) r
      (c:r) -> goCls (CChar c : acc) r
parseClass s = (Cls False [], s)

parseQuantSuffix :: Node -> String -> (Node, String)
parseQuantSuffix atom s = case s of
  ('*':rest) -> let (lz, r) = lazyq rest in (Star (not lz) atom, r)
  ('+':rest) -> let (lz, r) = lazyq rest in (Plus (not lz) atom, r)
  ('?':rest) -> let (lz, r) = lazyq rest in (Opt (not lz) atom, r)
  ('{':rest) ->
    let (mn, r1) = num rest
        (mx, r2) = case r1 of
          (',':r) -> let (s2, r') = num r in (if null s2 then Nothing else Just (read s2), r')
          _ -> (Just (if null mn then 0 else read mn), r1)
    in case r2 of
         ('}':r3) | not (null mn) ->
           let (lz, r4) = lazyq r3 in (Rep (not lz) (read mn) mx atom, r4)
         _ -> (atom, s)
  _ -> (atom, s)
  where
    lazyq ('?':r) = (True, r)
    lazyq r = (False, r)
    num = span (\c -> c >= '0' && c <= '9')

-- ----- matcher (backtracking, CPS over Maybe, capture-tracking) -----

isWord :: Char -> Bool
isWord c = (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z')
        || (c >= '0' && c <= '9') || c == '_'

spaceChars :: String
spaceChars = [' ', '\t', '\n', '\r', '\f', '\v']

citemMatch :: Citem -> Char -> Bool
citemMatch it c = case it of
  CChar x -> c == x
  CRange a b -> c >= a && c <= b
  CD -> c >= '0' && c <= '9'
  CND -> not (c >= '0' && c <= '9')
  CW -> isWord c
  CNW -> not (isWord c)
  CS -> c `elem` spaceChars
  CNS -> not (c `elem` spaceChars)

-- Capture spans as an association list; a group re-recorded on a later
-- iteration is simply prepended, so `lookup` sees the most recent span.
-- Spans are threaded functionally, so backtracking naturally discards
-- abandoned group records.
type Caps = [(Int, (Int, Int))]

mNode :: (Int -> Char) -> Int -> Node -> Int -> Caps -> (Int -> Caps -> Maybe r) -> Maybe r
mNode inp len node pos caps k = case node of
  Char c -> if pos < len && inp pos == c then k (pos + 1) caps else Nothing
  Any -> if pos < len && inp pos /= '\n' then k (pos + 1) caps else Nothing
  Start -> if pos == 0 then k pos caps else Nothing
  End -> if pos == len then k pos caps else Nothing
  WordB ->
    let before = pos > 0 && isWord (inp (pos - 1))
        after = pos < len && isWord (inp pos)
    in if before /= after then k pos caps else Nothing
  NWordB ->
    let before = pos > 0 && isWord (inp (pos - 1))
        after = pos < len && isWord (inp pos)
    in if before == after then k pos caps else Nothing
  Cls neg items ->
    if pos < len
      then let c = inp pos
               hit = any (`citemMatch` c) items
           in if (if neg then not hit else hit) then k (pos + 1) caps else Nothing
      else Nothing
  Grp gi alts ->
    let k' = if gi == 0 then k
             else \p caps' -> k p ((gi, (pos, p)) : caps')
    in asum [mSeq inp len sq pos caps k' | sq <- alts]
  Opt greedy a ->
    if greedy then mNode inp len a pos caps k <|> k pos caps
    else k pos caps <|> mNode inp len a pos caps k
  Star greedy a -> mStar inp len greedy a pos caps k
  Plus greedy a -> mNode inp len a pos caps (\p c -> mStar inp len greedy a p c k)
  Rep greedy mn mx a -> mRep inp len greedy mn mx a pos caps k

mStar :: (Int -> Char) -> Int -> Bool -> Node -> Int -> Caps -> (Int -> Caps -> Maybe r) -> Maybe r
mStar inp len greedy a pos caps k =
  let more p c = if p > pos then mStar inp len greedy a p c k else Nothing
  in if greedy
    then mNode inp len a pos caps more <|> k pos caps
    else k pos caps <|> mNode inp len a pos caps more

mRep :: (Int -> Char) -> Int -> Bool -> Int -> Maybe Int -> Node -> Int -> Caps -> (Int -> Caps -> Maybe r) -> Maybe r
mRep inp len greedy mn mx a pos caps k =
  if mn > 0
    then mNode inp len a pos caps (\p c -> mRep inp len greedy (mn - 1) (fmap (subtract 1) mx) a p c k)
    else case mx of
      Just 0 -> k pos caps
      _ ->
        let next p c = if p > pos then mRep inp len greedy 0 (fmap (subtract 1) mx) a p c k else Nothing
        in if greedy then mNode inp len a pos caps next <|> k pos caps
           else k pos caps <|> mNode inp len a pos caps next

mSeq :: (Int -> Char) -> Int -> [Node] -> Int -> Caps -> (Int -> Caps -> Maybe r) -> Maybe r
mSeq inp len sq pos caps k = case sq of
  [] -> k pos caps
  (x:rest) -> mNode inp len x pos caps (\p c -> mSeq inp len rest p c k)

compile :: String -> Re
compile = parse

mkInp :: String -> (Int -> Char, Int)
mkInp input =
  let len = length input
      arr = listArray (0, len - 1) input :: Array Int Char
  in ((arr !), len)

-- First match at or after `start`: (mstart, mend, caps).
findAt :: Re -> (Int -> Char) -> Int -> Int -> Maybe (Int, Int, Caps)
findAt re inp len = tryAt
  where
    tryAt i
      | i > len = Nothing
      | otherwise = case asum [mSeq inp len sq i [] (\p c -> Just (p, c)) | sq <- reAlts re] of
          Just (e, caps) -> Just (i, e, caps)
          Nothing -> tryAt (i + 1)

-- Does the pattern match anywhere in input?
test :: Re -> String -> Bool
test re input = let (inp, len) = mkInp input in isJust (findAt re inp len 0)

testStr :: String -> String -> Bool
testStr pat input = test (compile pat) input

-- Leftmost match: returns (start, stop) or Nothing.
findBounds :: Re -> String -> Maybe (Int, Int)
findBounds re input =
  let (inp, len) = mkInp input
  in fmap (\(s, e, _) -> (s, e)) (findAt re inp len 0)

subStr :: String -> Int -> Int -> String
subStr input a b = take (b - a) (drop a input)

matchGroups :: Re -> String -> Int -> Int -> Caps -> [String]
matchGroups re input ms me caps =
  subStr input ms me
    : [ maybe "" (\(a, b) -> subStr input a b) (lookup gi caps)
      | gi <- [1 .. reNgroups re] ]

-- First match as [whole, capture1, ...]; unmatched groups are "".
find :: Re -> String -> Maybe [String]
find re input =
  let (inp, len) = mkInp input
  in fmap (\(ms, me, caps) -> matchGroups re input ms me caps) (findAt re inp len 0)

-- Every non-overlapping match, left to right (Go FindAllStringSubmatch).
findAll :: Re -> String -> [[String]]
findAll re input = go 0
  where
    (inp, len) = mkInp input
    go pos
      | pos > len = []
      | otherwise = case findAt re inp len pos of
          Nothing -> []
          Just (ms, me, caps) ->
            matchGroups re input ms me caps
              : go (if me == ms then me + 1 else me)

-- Replace every match; the template supports JS-style refs ($& for the whole
-- match, $1..$9 for captures, $$ for a literal $). On a zero-width match the
-- current character is emitted and the scan advances one position.
replaceAll :: Re -> String -> String -> String
replaceAll re input template = go 0
  where
    (inp, len) = mkInp input
    go pos
      | pos > len = ""
      | otherwise = case findAt re inp len pos of
          Nothing -> subStr input pos len
          Just (ms, me, caps) ->
            subStr input pos ms
              ++ expand ms me caps template
              ++ (if me == ms
                    then (if ms < len then [inp ms] else "") ++ go (ms + 1)
                    else go me)
    expand _ _ _ [] = []
    expand ms me caps ('$':'$':r) = '$' : expand ms me caps r
    expand ms me caps ('$':'&':r) = subStr input ms me ++ expand ms me caps r
    expand ms me caps ('$':d:r)
      | d >= '0' && d <= '9' =
          let gi = fromEnum d - fromEnum '0'
              part
                | gi == 0 = subStr input ms me
                | gi <= reNgroups re =
                    maybe "" (\(a, b) -> subStr input a b) (lookup gi caps)
                | otherwise = ""
          in part ++ expand ms me caps r
    expand ms me caps (c:r) = c : expand ms me caps r

File: ocaml/src/vregex.ml
(* Minimal backtracking regex engine for the OCaml port of voxgig/struct.
 * Supports the RE2 subset the corpus exercises: literals, '.', anchors ^ $,
 * word boundaries (b and B forms), character classes [..] / [^..] with ranges
 * and \d \w \s \D \W \S, groups (..), (?:..) and (?P<name>..), alternation |,
 * quantifiers * + ? and {n}/{n,}/{n,m} with optional lazy '?'. Capturing
 * groups are tracked, so the engine backs the full re_* API (find with
 * captures, find_all, replace) at the Go-stdlib minimum bar; `test` also
 * serves $LIKE. No third-party dependency. *)

type node =
  | Char of char
  | Any
  | Start
  | End
  | WordB
  | NWordB
  | Cls of bool * citem list           (* negated?, items *)
  | Grp of int * node list list        (* capture index (0 = non-capturing), alternation *)
  | Star of bool * node                (* greedy?, atom *)
  | Plus of bool * node
  | Opt of bool * node
  | Rep of bool * int * int option * node

and citem =
  | CChar of char
  | CRange of char * char
  | CD | CW | CS | CND | CNW | CNS     (* d w s D W S escapes *)

(* Compiled = the alternation AST plus the number of capturing groups. *)
type t = { alts : node list list; ngroups : int }

(* ----- parser ----- *)

let parse (pat : string) : t =
  let n = String.length pat in
  let pos = ref 0 in
  let gc = ref 0 in
  let peek () = if !pos < n then Some pat.[!pos] else None in
  let adv () = incr pos in
  let parse_class () =
    (* assumes current char is '[' *)
    adv ();
    let neg = (peek () = Some '^') in
    if neg then adv ();
    let items = ref [] in
    let finished = ref false in
    while not !finished do
      match peek () with
      | None -> finished := true
      | Some ']' -> adv (); finished := true
      | Some '\\' ->
        adv ();
        (match peek () with
         | Some 'd' -> items := CD :: !items; adv ()
         | Some 'w' -> items := CW :: !items; adv ()
         | Some 's' -> items := CS :: !items; adv ()
         | Some 'D' -> items := CND :: !items; adv ()
         | Some 'W' -> items := CNW :: !items; adv ()
         | Some 'S' -> items := CNS :: !items; adv ()
         | Some 'n' -> items := CChar '\n' :: !items; adv ()
         | Some 't' -> items := CChar '\t' :: !items; adv ()
         | Some 'r' -> items := CChar '\r' :: !items; adv ()
         | Some c -> items := CChar c :: !items; adv ()
         | None -> ())
      | Some c ->
        adv ();
        (* range? *)
        (match peek () with
         | Some '-' when (!pos + 1 < n && pat.[!pos + 1] <> ']') ->
           adv ();
           (match peek () with
            | Some c2 -> adv (); items := CRange (c, c2) :: !items
            | None -> items := CChar c :: !items)
         | _ -> items := CChar c :: !items)
    done;
    Cls (neg, List.rev !items)
  in
  let parse_quant_suffix atom =
    match peek () with
    | Some '*' -> adv ();
      let lazy_ = (peek () = Some '?') in if lazy_ then adv ();
      Some (Star (not lazy_, atom))
    | Some '+' -> adv ();
      let lazy_ = (peek () = Some '?') in if lazy_ then adv ();
      Some (Plus (not lazy_, atom))
    | Some '?' -> adv ();
      let lazy_ = (peek () = Some '?') in if lazy_ then adv ();
      Some (Opt (not lazy_, atom))
    | Some '{' ->
      (* {n} {n,} {n,m} *)
      let save = !pos in
      adv ();
      let num () =
        let b = Buffer.create 4 in
        let rec go () = match peek () with
          | Some c when c >= '0' && c <= '9' -> Buffer.add_char b c; adv (); go ()
          | _ -> () in
        go (); Buffer.contents b in
      let mn = num () in
      let mx =
        match peek () with
        | Some ',' -> adv (); let s = num () in if s = "" then None else Some (int_of_string s)
        | _ -> Some (if mn = "" then 0 else int_of_string mn)
      in
      (match peek () with
       | Some '}' when mn <> "" ->
         adv ();
         let lazy_ = (peek () = Some '?') in if lazy_ then adv ();
         Some (Rep (not lazy_, int_of_string mn, mx, atom))
       | _ -> pos := save; None)  (* not a valid quantifier; treat '{' literally *)
    | _ -> None
  in
  let rec parse_alt () : node list list =
    let first = parse_seq () in
    let alts = ref [first] in
    while peek () = Some '|' do
      adv ();
      alts := parse_seq () :: !alts
    done;
    List.rev !alts
  and parse_seq () : node list =
    let out = ref [] in
    let stop = ref false in
    while not !stop do
      match peek () with
      | None | Some '|' | Some ')' -> stop := true
      | _ ->
        let atom = parse_atom () in
        (match atom with
         | None -> stop := true
         | Some a ->
           let a = (match parse_quant_suffix a with Some q -> q | None -> a) in
           out := a :: !out)
    done;
    List.rev !out
  and parse_atom () : node option =
    match peek () with
    | None -> None
    | Some '(' ->
      adv ();
      (* (?: is non-capturing; a plain ( and an RE2 named group (?P<name> are
         both capturing (names are not tracked). Indices follow open parens. *)
      let gi =
        if peek () = Some '?' && !pos + 1 < n && pat.[!pos + 1] = ':' then (adv (); adv (); 0)
        else if peek () = Some '?' && !pos + 2 < n && pat.[!pos + 1] = 'P'
                && pat.[!pos + 2] = '<' then begin
          adv (); adv (); adv ();
          while peek () <> None && peek () <> Some '>' do adv () done;
          (if peek () = Some '>' then adv ());
          incr gc; !gc
        end
        else (incr gc; !gc)
      in
      let alts = parse_alt () in
      (if peek () = Some ')' then adv ());
      Some (Grp (gi, alts))
    | Some '[' -> Some (parse_class ())
    | Some '.' -> adv (); Some Any
    | Some '^' -> adv (); Some Start
    | Some '$' -> adv (); Some End
    | Some '\\' ->
      adv ();
      (match peek () with
       | Some 'd' -> adv (); Some (Cls (false, [CD]))
       | Some 'w' -> adv (); Some (Cls (false, [CW]))
       | Some 's' -> adv (); Some (Cls (false, [CS]))
       | Some 'D' -> adv (); Some (Cls (false, [CND]))
       | Some 'W' -> adv (); Some (Cls (false, [CNW]))
       | Some 'S' -> adv (); Some (Cls (false, [CNS]))
       | Some 'b' -> adv (); Some WordB
       | Some 'B' -> adv (); Some NWordB
       | Some 'n' -> adv (); Some (Char '\n')
       | Some 't' -> adv (); Some (Char '\t')
       | Some 'r' -> adv (); Some (Char '\r')
       | Some c -> adv (); Some (Char c)
       | None -> Some (Char '\\'))
    | Some c -> adv (); Some (Char c)
  in
  let alts = parse_alt () in
  { alts; ngroups = !gc }

(* ----- matcher (backtracking, CPS, capture-tracking) ----- *)

let is_word c =
  (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') || (c >= '0' && c <= '9') || c = '_'

let citem_match it c =
  match it with
  | CChar x -> c = x
  | CRange (a, b) -> c >= a && c <= b
  | CD -> c >= '0' && c <= '9'
  | CND -> not (c >= '0' && c <= '9')
  | CW -> is_word c
  | CNW -> not (is_word c)
  | CS -> c = ' ' || c = '\t' || c = '\n' || c = '\r' || c = '\012' || c = '\011'
  | CNS -> not (c = ' ' || c = '\t' || c = '\n' || c = '\r' || c = '\012' || c = '\011')

(* Capture spans, one slot per capturing group (group i at index i-1). The
   continuation receives the accepting end position and the spans so far; the
   first accepting continuation under backtracking order wins. Spans are
   threaded functionally (copy-on-write), so backtracking discards abandoned
   group records naturally. *)
type caps = (int * int) option array

let rec m_node input len node pos (caps : caps)
    (k : int -> caps -> (int * caps) option) : (int * caps) option =
  match node with
  | Char c -> if pos < len && input.[pos] = c then k (pos + 1) caps else None
  | Any -> if pos < len && input.[pos] <> '\n' then k (pos + 1) caps else None
  | Start -> if pos = 0 then k pos caps else None
  | End -> if pos = len then k pos caps else None
  | WordB ->
    let before = pos > 0 && is_word input.[pos - 1] in
    let after = pos < len && is_word input.[pos] in
    if before <> after then k pos caps else None
  | NWordB ->
    let before = pos > 0 && is_word input.[pos - 1] in
    let after = pos < len && is_word input.[pos] in
    if before = after then k pos caps else None
  | Cls (neg, items) ->
    if pos < len then begin
      let c = input.[pos] in
      let hit = List.exists (fun it -> citem_match it c) items in
      if (if neg then not hit else hit) then k (pos + 1) caps else None
    end else None
  | Grp (gi, alts) ->
    let k' =
      if gi = 0 then k
      else (fun p caps' ->
          let c2 = Array.copy caps' in
          c2.(gi - 1) <- Some (pos, p);
          k p c2)
    in
    first_alt input len alts pos caps k'
  | Opt (greedy, a) ->
    if greedy then
      (match m_node input len a pos caps k with Some r -> Some r | None -> k pos caps)
    else
      (match k pos caps with Some r -> Some r | None -> m_node input len a pos caps k)
  | Star (greedy, a) -> m_star input len greedy a pos caps k
  | Plus (greedy, a) ->
    m_node input len a pos caps (fun p c -> m_star input len greedy a p c k)
  | Rep (greedy, mn, mx, a) -> m_rep input len greedy mn mx a pos caps k

and first_alt input len alts pos caps k =
  match alts with
  | [] -> None
  | seq :: rest ->
    (match m_seq input len seq pos caps k with
     | Some r -> Some r
     | None -> first_alt input len rest pos caps k)

and m_star input len greedy a pos caps k =
  let more p c = if p > pos then m_star input len greedy a p c k else None in
  if greedy then
    (match m_node input len a pos caps more with Some r -> Some r | None -> k pos caps)
  else
    (match k pos caps with Some r -> Some r | None -> m_node input len a pos caps more)

and m_rep input len greedy mn mx a pos caps k =
  if mn > 0 then
    m_node input len a pos caps (fun p c ->
        m_rep input len greedy (mn - 1)
          (match mx with Some m -> Some (m - 1) | None -> None) a p c k)
  else
    match mx with
    | Some 0 -> k pos caps
    | _ ->
      let next p c = if p > pos then m_rep input len greedy 0
                       (match mx with Some m -> Some (m - 1) | None -> None) a p c k
        else None in
      if greedy then
        (match m_node input len a pos caps next with Some r -> Some r | None -> k pos caps)
      else
        (match k pos caps with Some r -> Some r | None -> m_node input len a pos caps next)

and m_seq input len seq pos caps k =
  match seq with
  | [] -> k pos caps
  | x :: rest -> m_node input len x pos caps (fun p c -> m_seq input len rest p c k)

let compile (pat : string) : t = parse pat

(* First match at or after `start`: (mstart, mend, caps). *)
let find_at (re : t) (input : string) (start : int) : (int * int * caps) option =
  let len = String.length input in
  let rec try_at i =
    if i > len then None
    else begin
      let caps0 = Array.make (max 1 re.ngroups) None in
      match first_alt input len re.alts i caps0 (fun p c -> Some (p, c)) with
      | Some (e, caps) -> Some (i, e, caps)
      | None -> try_at (i + 1)
    end
  in try_at start

(* Does the pattern match anywhere in input? *)
let test (re : t) (input : string) : bool =
  match find_at re input 0 with Some _ -> true | None -> false

let test_str (pat : string) (input : string) : bool = test (compile pat) input

(* Leftmost match: returns (start, stop) or None. *)
let find_bounds (re : t) (input : string) : (int * int) option =
  match find_at re input 0 with
  | Some (s, e, _) -> Some (s, e)
  | None -> None

let group_str input (caps : caps) gi =
  match caps.(gi) with
  | Some (a, b) -> String.sub input a (b - a)
  | None -> ""

let match_groups (re : t) input ms me caps =
  String.sub input ms (me - ms)
  :: List.init re.ngroups (fun i -> group_str input caps i)

(* First match as [whole; capture1; ...]; unmatched groups are "". *)
let find (re : t) (input : string) : string list option =
  match find_at re input 0 with
  | None -> None
  | Some (ms, me, caps) -> Some (match_groups re input ms me caps)

(* Every non-overlapping match, left to right (Go FindAllStringSubmatch). *)
let find_all (re : t) (input : string) : string list list =
  let len = String.length input in
  let rec go pos acc =
    if pos > len then List.rev acc
    else match find_at re input pos with
      | None -> List.rev acc
      | Some (ms, me, caps) ->
        let m = match_groups re input ms me caps in
        let pos' = if me = ms then me + 1 else me in
        go pos' (m :: acc)
  in
  go 0 []

(* Replace every match; the template supports the JS-style refs (and-sign for
   the whole match), 1..9 group refs and a doubled dollar for a literal
   dollar. On a zero-width match the current character is emitted and the
   scan advances one position. *)
let replace_all (re : t) (input : string) (template : string) : string =
  let len = String.length input in
  let b = Buffer.create (String.length input) in
  let tn = String.length template in
  let expand ms me (caps : caps) =
    let i = ref 0 in
    while !i < tn do
      let c = template.[!i] in
      if c = '$' && !i + 1 < tn then begin
        let nc = template.[!i + 1] in
        if nc = '$' then (Buffer.add_char b '$'; i := !i + 2)
        else if nc = '&' || (nc >= '0' && nc <= '9') then begin
          (if nc = '&' || nc = '0' then
             Buffer.add_string b (String.sub input ms (me - ms))
           else begin
             let gi = Char.code nc - Char.code '0' in
             if gi <= re.ngroups then
               (match caps.(gi - 1) with
                | Some (a, e) -> Buffer.add_string b (String.sub input a (e - a))
                | None -> ())
           end);
          i := !i + 2
        end else (Buffer.add_char b c; incr i)
      end else (Buffer.add_char b c; incr i)
    done
  in
  let rec go pos =
    if pos > len then ()
    else match find_at re input pos with
      | None -> Buffer.add_string b (String.sub input pos (len - pos))
      | Some (ms, me, caps) ->
        Buffer.add_string b (String.sub input pos (ms - pos));
        expand ms me caps;
        if me = ms then begin
          (if ms < len then Buffer.add_char b input.[ms]);
          go (ms + 1)
        end else go me
  in
  go 0;
  Buffer.contents b
